paperqa/readers.py

```diff
@@ -152,25 +206,80 @@ def parse_code_txt(path: Path, doc: Doc, chunk_chars: int, overlap: int) -> list
     return texts
 
 
-def read_doc(
+def read_doc(  # noqa: PLR0912
     path: Path,
     doc: Doc,
     chunk_chars: int = 3000,
     overlap: int = 100,
     force_pypdf: bool = False,
-) -> list[Text]:
-    """Parse a document into chunks."""
+    parsed_text_only: bool = False,
+    include_metadata: bool = False,
+) -> list[Text] | ParsedText | tuple[list[Text], ParsedMetadata]:
```

Comment:
(per discussion in-person)

This complex return type arises from multiple `bool` configuration arguments, which will lead to installers of `paper-qa` (and within `paper-qa` itself) having to add `type: ignore` comments. An immediate (yet verbose) solution is using `typing.overload` to better express this signature.

However, in general, this complex signature is a code smell for a decomposition needing to take place. Maybe in the future we can decompose this better to make the code more fluent without requiring big signatures

Reply:
Added! Was able to remove the type ignores from `docs.py` after implementing.

Reply:
https://youtu.be/E5KfzyO4feo?t=21

@mskarlin is now Tron at 0:21 after making this change